plenum/server/consensus/view_change_service.py

```diff
@@ -1,30 +1,113 @@
 from _sha256 import sha256
-from typing import List
+from collections import defaultdict
+from functools import partial
+from typing import List, Dict, Optional, Union
 
 from common.serializers.json_serializer import JsonSerializer
 from plenum.common.event_bus import InternalBus, ExternalBus
 from plenum.common.messages.node_messages import ViewChange, ViewChangeAck, NewView
+from plenum.common.stashing_router import StashingRouter, PROCESS, DISCARD, STASH
 from plenum.common.timer import TimerService
 from plenum.server.consensus.consensus_data_provider import ConsensusDataProvider
 from plenum.server.quorums import Quorums
 
 
+def view_change_digest(msg: ViewChange) -> str:
+    serialized = JsonSerializer().dumps(msg.__dict__)
+    return sha256(serialized).hexdigest()
+
+
+class ViewChangeVotesForNode:
+    """
+    Storage for view change vote from some node for some view + corresponding acks
+    """
+
+    def __init__(self, quorums: Quorums):
+        self._quorums = quorums
+        self._view_change = None
+        self._digest = None
+        self._acks = defaultdict(set)  # Dict[str, Set[str]]
+
+    @property
+    def view_change(self) -> Optional[ViewChange]:
+        """
+        Returns received view change
+        """
+        return self._view_change
+
+    @property
+    def is_confirmed(self) -> bool:
+        """
+        Returns True if received view change message and enough corresponding acks
+        """
+        if self._digest is None:
+            return False
+
+        return self._quorums.view_change_ack.is_reached(len(self._acks[self._digest]))
+
+    def add_view_change(self, msg: ViewChange) -> bool:
```

Comment:
How are we going to use the return value?

Reply:
I was planning to raise suspicion if it returned false. Probably it should return either tuple(ok: bool, err: str) or just string.